[clang][bytecode] Start implementing __builtin_bit_cast

tbaederr opened the pull request:
This is a subset of #68288, with hopefully narrower scope. It does not support bitcasting to non-integral types yet.
Bitfields are supported, but only if they result in a full byte-sized final buffer. It does not support casting from null-pointers yet or casting from indeterminate bits.


The tests are from #68288 and partially from #74775.

The `BitcastBuffer` struct is currently always working in single bits, but I plan to (try to) optimize this for the common full-byte case.

tbaederr committed:
[clang][bytecode] Start implementing __builtin_bit_cast

erichkeane commented on clang/lib/AST/ByteCode/Boolean.h:
```diff
@@ -81,6 +81,14 @@ class Boolean final {
   Boolean truncate(unsigned TruncBits) const { return *this; }
 
+  static Boolean bitcastFromMemory(const std::byte *Buff, unsigned BitWidth) {
+    assert(BitWidth == 8);
```
Perhaps a `//FIXME: <why this is limited to 8 bits, and what we have to do to fix this>` here.

erichkeane commented on clang/lib/AST/ByteCode/Compiler.cpp:
```diff
@@ -6400,6 +6403,67 @@ bool Compiler<Emitter>::emitDestruction(const Descriptor *Desc,
+  QualType ToType = E->getType();
+  std::optional<PrimType> ToT = classify(ToType);
+
+  assert(!DiscardResult && "Implement");
```
would appreciate more detail here, just a little more info on the assert message itself.

erichkeane commented on clang/lib/AST/ByteCode/Compiler.cpp:
```diff
@@ -6400,6 +6403,67 @@ bool Compiler<Emitter>::emitDestruction(const Descriptor *Desc,
+
+  if (!ToT || ToT == PT_Ptr) {
+    // Conversion to an array or record type.
+    assert(false && "Implement");
```
same here.

erichkeane reviewed (commented):
Implementation seems reasonable to me, a couple of nits (and please make more detailed asserts for everything!), though I haven't paid attention to these reviews in long enough someone else should take a look.

tbaederr commented:
Marking this as a draft until I have fixed things on big-endian hosts.

tbaederr commented:
Victory!

 1) Iterate composite fields backwards for big-endian targets
 2) Fix `BitcatBuffer::data()` pointing to the wrong byte if the number of bytes in the buffer is not a multiple of `sizeof(uint64_t)` (which is what `llvm::BitBuffer` uses internally) and we're on a big-endian host.
 3) Validate that all of this works on little-endian and big-endian hosts, with asan and ubsan enabled.

tbaederr commented:
Also, I've removed the logic for tracking uninitialized/indeterminate bits for now since it can be added later.

tbaederr committed:
Add test for casting to nullptr_t

tbaederr committed:
Fix precommit CI failure

AaronBallman commented on clang/lib/AST/ByteCode/Integral.h:
```diff
@@ -154,6 +154,18 @@ template <unsigned Bits, bool Signed> class Integral final {
+    assert(BitWidth == sizeof(ReprT) * 8);
+    ReprT V;
+
+    std::memcpy(&V, Src, sizeof(ReprT));
```
Can we add an assert somewhere that `ReprT` will always be trivially copyable?

AaronBallman commented on clang/test/AST/ByteCode/builtin-bit-cast.cpp:
```diff
@@ -0,0 +1,399 @@
+  };
+  // expected-error@+2 {{constexpr variable 'g' must be initialized by a constant expression}}
+  // expected-note@+1 {{bit_cast from a union type is not allowed in a constant expression}}
+  constexpr G g = __builtin_bit_cast(G, X{0});
```
Why is this test disabled? I would have assumed this worked?

tbaederr replied:
It's bitcasting _to_ a non-primitive type, that's not part of this PR.

AaronBallman replied:
Ah! That makes sense -- can you add a test that shows bitcasting from a union to a primitive type is caught? e.g.,
```
constexpr int a = __builtin_bit_cast(int, X{0});
```

AaronBallman reviewed (commented):
This generally LGTM, but I'd like to see some tests for slightly more complicated situations, such as derived classes, classes with reference members, bitcasting a class with a virtual function, bitcasting a class with a virtual base, etc.

tbaederr committed:
Address review comments

 1) add the static_assert
 2) Add more tests
 3) for 2), Fix iterating over record bases/fields. need to
    invert the field/base order as well.

tbaederr commented:
> such as derived classes,

Got me again :disappointed:

tbaederr committed:
Add a test with a class with 2 bases

AaronBallman reviewed (approved):
LGTM with a bit of extra test coverage

tbaederr committed:
Add  a union -> int test case

tbaederr commented:
Just tested this on a big-endian host again, no problems found.
